MapBox: get access to feature coordinates for parametrable style.base_alti

ftoromanoff opened the pull request:
Currently Vector Tile style is defined using the MapBox Stylesheet enter in the VectorTileSource and there is no way to modify it outside of the itowns code.

This PR aims to be able to add dynamic function allowing the use of the feature altitude as an argument.

Other modifications:
- reorganization of code lines in Feature2Mesh.js to be similar between all featureTo[Point/Line/Polygon/ExtrudedPolygon] functions
- - add a new example using the MapBox Streets v8 tileSet flux
- change order of arguments in feature.pushCoordinates() 
-> to homogenize arg order of feature.pushCoordinates() and feature.pushCoordinatesValues() 

Breaking changes: 
- feature.pushCordinates(): change order of arguments.
- feature.pushCordinatesValues(): change in signature.

ftoromanoff committed:
feat(VectorTile): add coordProj in pushCoordinatesValues() to get access to position when using style.base_altitude.

BREAKING CHANGE: Feature.pushCoordinatesValues() signature change from pushCoordinatesValues(feature, long, lat, normal) to pushCoordinatesValues(feature, coordIn, coordProj)

ftoromanoff committed:
example(VectorTile): new example using official mapbox flux, showing buildings placed at ground level.

jailln commented on examples/vector_tile_3d_mesh_Mapbox.html:
```diff
@@ -0,0 +1,131 @@
+                view.addLayer(new itowns.ElevationLayer(config.id, config))
+                    .then(debugMenu.addLayerGUI.bind(debugMenu));
+            }
+            // itowns.Fetcher.json('./layers/JSONLayers/IGN_MNT_HIGHRES.json').then(addElevationLayerFromConfig);
```
remove

ftoromanoff replied:
done

jailln commented on examples/vector_tile_3d_mesh_Mapbox.html:
```diff
@@ -0,0 +1,131 @@
+
+            // // Define the source of the building data : those are vector tiled data from the geoportail.
+            const buildingsSource = new itowns.VectorTilesSource({
+                style: 'http://localhost:8081/testMvt.json',
```
Can you publish this service or use a published one ?

ftoromanoff replied:
done

jailln commented on examples/vector_tile_3d_mesh_Mapbox.html:
```diff
@@ -0,0 +1,131 @@
+                source: buildingsSource,
+                zoom: {
+                    min: 15,
+                    // max: 13,
```
remove

ftoromanoff replied:
done

jailln commented on examples/vector_tile_3d_mesh_Mapbox.html:
```diff
@@ -0,0 +1,131 @@
+                style: {
+                    fill: {
+                        base_altitude: (p, c) => {
+                            return (altitudeBuildings(p, c) || 0) + p.min_height;
```
this computation can be done in `altitudeBuildings` directly

ftoromanoff replied:
I separate both to show more easily the part where we get the ground elevation data for a given building. The p.min_height is a specificité of the mapbow streets V8 building tile set (for building parts that are not directly on the ground).

jailln replied:
ok, I found it clearer to put everything in one function but it's not critical so I'll let you decide

jailln commented on src/Parser/VectorTileParser.js:
```diff
@@ -4,12 +4,25 @@ import { VectorTile } from '@mapbox/vector-tile';
 const lastPoint = new Vector2();
 const firstPoint = new Vector2();
 
+function project(x, y, tileNumbers, tileExtent) {
```
Can you add some comments to avoid having to get into the maths to understand what is projected on what each time we come accross this function plase ?:)

ftoromanoff replied:
I added a little comment to say from where i got the function.

jailln commented on src/Process/FeatureProcessing.js:
```diff
@@ -34,6 +34,8 @@ export default {
         const zoomDest = extentsDestination[0].zoom;
 
         // check if it's tile level is equal to display level layer.
+        // TO DO updata at all level asked
```
can you create an issue please? (or add it to the issue on vector data you already created)

ftoromanoff replied:
It's the first item in my issue on vector tile. I tried to clarified it on my list.

jailln commented on src/Core/Feature.js:
```diff
@@ -135,14 +135,13 @@ export class FeatureGeometry {
+     * @param {Coordinates} coordIn The coordinates to push.
      */
-    pushCoordinates(coordIn, feature) {
+    pushCoordinates(feature, coordIn) {
```
this change seems not necessary and it breaks the API, I would prefer that we dont do it.

ftoromanoff replied:
I did this change for the 2 methods: pushCoordinates() and pushCoordinatesValues() to be more similaire as both do the same thing, the second one had been added for the unique case of vectorTile support. The purpose at he end is to have only 1 methode to do both.

jailln replied:
Ok, can't we go all the way and keep only one without modifying its argument list ? and deprecate the other (by calling the new one) ; so we don't have a breaking change

ftoromanoff replied:
It would be the topic of a new PR aiming at clearing all the process of creating new itowns.Features I think.
It's in the idea mostly link with the way how is handle base_altitude.

jailln replied:
I think it would be interresting to do at least what I propose in my previous comment in this PR so we don't break the API.

ftoromanoff replied:
I added a deprecation message. It should avoid Breaking the API.

jailln replied:
Thanks! Can you do the same for `pushCoordinatesValues` please?

jailln replied:
Also, can you add `BREAKING CHANGE: <description of the changes>` in the commits that modify these methods please?

ftoromanoff replied:
1) I didn't change the way pushCoordinatesValues() is called, there is no deprecation there.
2) I did add the BREAKING CHANGE message but indeed not in the good commit. I will correct that mistake now.

jailln replied:
1. From what I see you moved its signature from `pushCoordinatesValues(feature, long, lat, normal)` to `pushCoordinatesValues(feature, coordIn, coordProj)`

jailln commented on src/Core/Feature.js:
```diff
@@ -154,23 +153,25 @@ export class FeatureGeometry {
+    pushCoordinatesValues(feature, coordIn, coordProj) {
+        const altitude = this.baseAltitude(feature, coordProj);
+
+        _setGeometryValues(this, feature, coordIn.x, coordIn.y, altitude, coordIn.normal);
```
shouldn't we also modify `_setGeometryValues` to take the following arguments:
`_setGeometryValues(this, feature, coordIn, altitude)` and let it get `coordIn.x`, `coordIn.y` and `coordIn.normal` when it needs it?

ftoromanoff replied:
It might be an idea, But i don't know if it's needed here.

jailln replied:
It's not mandatory but it would make sense to do it now in my opinion since you started simplifying `pushCoordinatesValues`

ftoromanoff replied:
Done

jailln reviewed (commented):
* rename example to `vector_tile_3d_mesh_Mapbox` to `vector_tiles_3d_mesh_mapbox`
* add the example to eamples config (`examples/config.json`)
* add breaking changes to you commit message (the modifications you made to the `Feature` API)

ftoromanoff committed:
refactor(VectorTile): Code clean up

ftoromanoff committed:
refactor(Feature): Feature.pushCoordinates() change in arguments order

BREAKING CHANGE: Feature.pushCoordinates() change in arguments order from pushCoordinates(coordIn, feature) to pushCoordinates(feature, coordIn)

jailln commented on test/functional/vector_tile_3d_mesh_mapbox.js:
```diff
@@ -0,0 +1,26 @@
+        assert.ok(result);
+    });
+
+    it('load features data', async function _2() {
```
rename to `should correctly load features at given coordinates`

ftoromanoff replied:
Corrected

jailln commented on src/Core/Feature.js:
```diff
@@ -154,23 +157,25 @@ export class FeatureGeometry {
-
-        _setGeometryValues(this, feature, long, lat, altitude, normal);
+     * @param {Object} coordIn An object containing the coordinates values to push.
+     * @param {number} coordIn.x the x coordinate (in a local systeme).
```
`system` without "e" at the end (same for `coordIn.y`)

mgermerie reviewed (approved):
I made a small update on examples comments. Thanks for the PR !
